torch/distributed/_shard/api.py

```diff
@@ -118,3 +123,80 @@ def shard_parameter(
 
     # Now we can set the attribute appropriately.
     setattr(module, param_name, st)
+
+
+class ShardedModuleSwapper(abc.ABC):
+    @abc.abstractmethod
+    def process(self, module: nn.Module) -> nn.Module:
+        """
+        Processes a module and if needed swaps it with a custom sharded
+        Implementation. Should return ``None`` if no swapping should be
+        performed.
+
+        The passed in module would consist of ShardedTensors and a common way
+        to perform module swapping would be to use the state_dict of the passed
+        in module and apply it to the new sharded module via its
+        load_state_dict method.
+        """
+        pass
+
+def shard_module(
+    module: nn.Module,
+    plan: ShardingPlan,
+    sharded_module_swapper: ShardedModuleSwapper = None,
+    reshard_output: ShardingPlan = None,
+    src_rank=0,
+    process_group=None
+):
+    """
+    Shards a given module according to the provided sharding_plan. This method
+    first shards all the parameters according to the given sharding_plan. Next,
+    If sharded_module_swapper is specified, it recursively traverses the module
+    tree and calls ``sharded_module_swapper.process()`` for each submodule.
+    If ``sharded_module_swapper.process()`` returns ``None``, the method continues
+    to recurse further down the tree. If ``sharded_module_swapper.process()``
+    returns an ``nn.Module``, then the current module is replaced with the return
+    value and we don't recurse further.
+
+    If ``reshard_output`` is specified, the module's output is resharded
+    according to the provided sharding plan.
+
+    Args:
+        module: The module to apply sharding to
+        sharding_plan: Dict from param name to PlacementSpec/ShardingSpec
```

Comment:
This is not a Dict but a `ShardingPlan` type? Shouldn't we include types in parenthesis as part of the dos?